examples/tut-6-rpc/client-go/index.go

```diff
@@ -0,0 +1,214 @@
+
+    res.Body.Close();
+    dieIf(json.Unmarshal(body, &ans))
+
+    fmt.Printf("%s ==> %s\n", lab, string(body))
+
+    return ans
+  }
+
+  rpcCallbacks := func(m string, arg contract, cbacks map[string]interface{}) {
+    var ans interface{}
+
+    vals  := make(map[string]interface{})
+    meths := make(map[string]bool)
+
+    for key, val := range cbacks {
+      if reflect.ValueOf(val).Kind() == reflect.Func {
+        meths[key] = true
+      } else {
+        vals[key]  = val
+      }
+    }
+
+    p := rpc(m, arg, vals, meths).(map[string]interface{})
+
+    for true {
+      if p["t"] == "Done" {
+        break
+
+      } else if p["t"] == "Kont" {
+
+        // Apply interact method to supplied inputs
+        if _, exists := meths[p["m"].(string)]; exists {
+          cb    := reflect.ValueOf(cbacks[p["m"].(string)])
+          args  := p["args"].([]interface{})
+          vargs := make([]reflect.Value, len(args))
+
+          for i, a := range args {
+            vargs[i] = reflect.ValueOf(a)
+          }
+
+          ans = nil
+          for _, a := range cb.Call(vargs) {
+            ans = a.Interface()
+            break
+          }
+
+          p = rpc("/kont", p["kid"], ans).(map[string]interface{})
+
+        // Respond with simple field value
+        } else {
+          ans = vals[p["m"].(string)]
+          p   = rpc("/kont", p["kid"], ans).(map[string]interface{})
+        }
+
+      } else {
+        log.Fatalf("Illegal callback return: %s", p)
+        os.Exit(1)
+      }
+    }
+  }
+
+  return rpc, rpcCallbacks
+}
+
+
+func main() {
+  fmt.Println("I am the client")
+  rpc, rpcCallbacks := mkRpc()
+
+  fmtc := func(i bigNumber) string {
+    return rpc("/stdlib/formatCurrency", i, 4).(string)
+  }
+
+  getBalance := func(w account) string {
+    return fmtc(rpc("/stdlib/balanceOf", w).(bigNumber))
+  }
+
+  startingBalance := rpc("/stdlib/parseCurrency", 10).(bigNumber)
+  accAlice        := rpc("/stdlib/newTestAccount", startingBalance).(account)
+  accBob          := rpc("/stdlib/newTestAccount", startingBalance).(account)
+
+  beforeAlice     := getBalance(accAlice)
+  beforeBob       := getBalance(accBob)
+
+  ctcAlice        := rpc("/acc/deploy",  accAlice).(contract)
+  aliceInfo       := rpc("/ctc/getInfo", ctcAlice).(interface{})
+  ctcBob          := rpc("/acc/attach",  accBob, aliceInfo).(contract)
+
+  HAND            := [3]string{"Rock", "Paper", "Scissors"}
+  OUTCOME         := [3]string{"Bob wins", "Draw", "Alice wins"}
+
+  player := func(who string) map[string]interface{} {
+    getHand := func() int {
+      // https://golang.org/pkg/math/rand/#example_Intn
+      rand.Seed(time.Now().UnixNano())
+      hand := rand.Intn(3)
+      fmt.Printf("%s played %s\n", who, HAND[hand])
+      return hand
+    }
+
+    informTimeout := func() {
+      fmt.Printf("%s observed a timeout\n", who)
+    }
+
+    seeOutcome := func(n bigNumber) {
+      o := int(rpc("/stdlib/bigNumberToNumber", n).(float64))
```

Comment:
Why `float64`?

Reply:
I'm guessing the type is inferred somewhere in the HTTP library's guts - there's a runtime exception if I don't capture it this way and cast to `int`.

Reply:
Ah